core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -2941,6 +2941,19 @@ private void registerQuery(
           SqlSelect.HAVING_OPERAND);
       registerSubQueries(selectScope2,
           SqlNonNullableAccessors.getSelectList(select));
+
+      if (enclosingNode.getKind() == SqlKind.UPDATE) {
+        registerSubQueries(selectScope2,
+            ((SqlUpdate) enclosingNode).getSourceExpressionList());
+      } else if (enclosingNode.getKind() == SqlKind.MERGE) {
+        SqlUpdate updateCall = ((SqlMerge) enclosingNode).getUpdateCall();
```

Comment:
this looks wrong, you mean SqlMerge?

Reply:
:thinking: what exactly looks wrong?
Here we register subqueries that may be present in updateCall (SqlUpdate), which may be present in SqlMerge statement.

Reply:
Sorry, my mistake.